EXTENSIONS.md

```diff
@@ -0,0 +1,121 @@
+# Linkerd Extensions
+
+Linkerd has an extension model which allows 3rd parties to add functionality.
+Each extension consists of a binary CLI named `linkerd-name` (where `name` is
+the name of the extension) as well as a set of Kubernetes resources.  To invoke
+an extension, users can run `linkerd name` which will search the current PATH
+for an executable named `linkerd-name` and execute it.
+
+## Installing Extensions
+
+To install an extension, first download the extension executable and put it
+in your PATH. The extension can then be installed into your Kubernetes cluster
+by running `linkerd <extension name> install | kubectl apply -f -`. Similarly,
+it can be uninstalled by running
+`linkerd <extension name> uninstall | kubectl delete -f -`.
+
+A full list of installed extensions can be printed by running `linkerd check`.
+
+## Developing Extensions
+
+The extension must be an executable file named `linkerd-name` where `name` is
+the name of the extension. The name must not be any of the built-in Linkerd
+commands (e.g. `check`) or extensions (e.g. `viz`).
+
+The extension must accept the following flags and respect them any time that
+it communicates with the Kubernetes API.  All of these flags must be accepted
+but may be ignored if they are not applicable.
+
+* `--api-addr`: Override kubeconfig and communicate directly with the control
+  plane at host:port (mostly for testing)
+* `--context`: Name of the kubeconfig context to use
+* `--as`: Username to impersonate for Kubernetes operations
+* `--as-group`: Group to impersonate for Kubernetes operations
+* `--help`/`-h`: Print help message
+* `--kubeconfig`: Path to the kubeconfig file to use for CLI requests
+* `--linkerd-namespace`/`-l`: Namespace in which Linkerd is installed
+  [$LINKERD_NAMESPACE]
+* `--verbose`: Turn on debug logging
+
+The extension must implement these commands:
+
+### `linkerd-name install`
+
+This command must print the Kubernetes manifests for the extension as yaml
+which is suitable to be passed to `kubectl apply -f`. These manifests must
+include a Namespace resource with the label `linkerd.io/extension=name` where
+`name` is the name of the extension. This allows Linkerd to detect installed
+extensions.
+
+### `linkerd-name uninstall`
+
+This command must print manifests for all Kubernetes resources belonging to
+this extension as yaml which is suitable to be passed to `kubectl delete -f`.
+Special care should be taken to make sure both namespace-scoped and cluster-
+scoped resources are listed.
+
+### `linkerd-name check`
+
+This command must perform any appropriate health checks for the extension
+including but not limited to checking that the extension resources exist and
+are in a healthy state. This command must exit with a status code of 0 if the
+checks pass or with a nonzero status code if they do not pass. For consistency,
+it is recommended that this command follows the same output formatting as
+`linkerd check`, e.g.
```

Comment:
As mentioned [here](https://github.com/linkerd/linkerd2/pull/5762#issuecomment-781470004), I think this might be too simplistic. I think that `linkerd` should ultimately be responsible for output formatting -- probably by _always_ specifying `--output=json` -- and the extension CLI should only be responsible for returning structured data to the main cli process.